src/System.Private.CoreLib/shared/System/Globalization/TextInfo.cs

```diff
@@ -373,49 +383,142 @@ private unsafe string ChangeCase(string source, bool toUpper)
             return result;
         }
 
-        internal unsafe void ChangeCase(ReadOnlySpan<char> source, Span<char> destination, bool toUpper)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ChangeCaseToLower(ReadOnlySpan<char> source, Span<char> destination)
         {
-            Debug.Assert(!_invariantMode);
             Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<ToLowerConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ChangeCaseToUpper(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<ToUpperConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ChangeCaseCommon<TConversion>(ReadOnlySpan<char> source, Span<char> destination) where TConversion : struct
+        {
+            Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<TConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        private void ChangeCaseCommon<TConversion>(ref char source, ref char destination, int charCount) where TConversion : struct
+        {
+            Debug.Assert(typeof(TConversion) == typeof(ToUpperConversion) || typeof(TConversion) == typeof(ToLowerConversion));
+            bool toUpper = typeof(TConversion) == typeof(ToUpperConversion); // JIT will treat this as a constant in release builds
 
-            if (source.IsEmpty)
+            Debug.Assert(!_invariantMode);
+            Debug.Assert(charCount >= 0);
+
+            if (charCount == 0)
             {
-                return;
+                goto Return;
             }
 
-            fixed (char* pSource = &MemoryMarshal.GetReference(source))
-            fixed (char* pResult = &MemoryMarshal.GetReference(destination))
+            if (IsAsciiCasingSameAsInvariant)
```

Comment:
If there's a desire to do so, we can refactor this "Boolean that might not be initialized" logic out into a standalone type. In an ideal world, the codegen would look like the following.

```asm
cmp byte ptr [ptr_to_field], 1
jl LABEL_NotTrue
LABEL_True:
.. logic where value evaluated to true ..

LABEL_NotTrue:
cmp byte ptr [ptr_to_field], 0
jl LABEL_NeedToEvaluate
LABEL_False:
.. logic where value evaluated to false ..

LABEL_NeedToEvaluate:
call Evaluate
cmp byte ptr [ptr_to_field], 0
ja LABEL_True
jmp LABEL_False
```

This means that in the common case of the value being already-evaluated and equal to _true_, no jumps at all are taken.
In the less common case of the value being already-evaluated and equal to _false_, only a single jcc is taken.
And in the uncommon case of the value not yet being evaluated, multiple jumps are taken.